### PR TITLE
20591-WideCharacterSet-is-not-a-subclass-of-CharacterSet

### PR DESCRIPTION
**Pull request by pavel-krivanek**
https://pharo.fogbugz.com/f/cases/20591/WideCharacterSet-is-not-a-subclass-of-CharacterSet

convert from https://github.com/pharo-project/pharo/pull/399/files

### CODE REVIEW

**Comment by pavel-krivanek** [comment]
This PR is not ready to be integrated. Just conversion of the old pre-tonel PR

**Comment by nicolas-cellier-aka-nice** [comment]
I find the bug report completely useless.

It does not tell what is the problem of old implementation, or what enhancement we could expect from a new implementation.
There is absolutely no rationalization, nor reference to an eventual place where this rationalization could have happened (mailing list, discussion channel, private conversation ???).

Sorry to be rude, but it's degree 0 of quality...

The changes by themselves might be good, but lenghty and difficult to review...
Maybe we would want to pick some of the changes and reject some others (like = which is questionnable).

From what i understand, the main benefits are the ability to perform union: and intersection: of (Wide)CharacterSet. Are there other benefits?

File: src/Glamour-Tests-Morphic/GLMWatcherMorphicTest.class.st [hunk 1]
```diff
@@ -4,12 +4,6 @@ Class {
 	#category : #'Glamour-Tests-Morphic'
 }
 
-{ #category : #tests }
```
Comment (nicolas-cellier-aka-nice) [inline src/Glamour-Tests-Morphic/GLMWatcherMorphicTest.class.st]: All changes below this one are unrelated and should be reverted

File: src/Collections-Support/WideCharacterSet.class.st [hunk 2]
```diff
@@ -18,14 +18,19 @@ A byteArrayMap is maintained in parallel with map for fast handling of ByteStrin
 "
 Class {
 	#name : #WideCharacterSet,
-	#superclass : #Collection,
+	#superclass : #CharacterSet,
 	#instVars : [
-		'map',
```
Comment (nicolas-cellier-aka-nice) [inline src/Collections-Support/WideCharacterSet.class.st]: super map is already a byteArrayMap...
Having to change class/semantic of a super instance is a smell...

So the change should better be:
- rename super map into byteArrayMap
- rename map into wideCharacterMap
From a bootstrap perspective, this may require several commits though...
(there were live instances to mutate into the Squeak image, so probably so in Pharo!)

File: src/Collections-Support/WideCharacterSet.class.st [hunk 3]
```diff
@@ -36,15 +41,24 @@ WideCharacterSet class >> newFrom: aCollection [
 
 { #category : #comparing }
 WideCharacterSet >> = anObject [
-	^self species == anObject species and: [
-		self wideCharacterMap = anObject wideCharacterMap ]
+	self class == anObject class
+		ifTrue: [
+			^ (self byteArrayMap = anObject byteArrayMap) and: [ 
+				self wideCharacterMap = anObject wideCharacterMap ] ].
+	((anObject respondsTo: #size) and: [ anObject respondsTo: #includes: ])
```
Comment (nicolas-cellier-aka-nice) [inline src/Collections-Support/WideCharacterSet.class.st]: Beware, a CharacterSetComplement respondsTo: #size... with shouldNotImplement
(until we fix an upper bound to the set of characters)

**Comment by Ducasse** [comment]
More work needed so let us close this pull request.
